### PR TITLE
CDPCP-1560. Allow retrieving keytab for different users

### PR DESCRIPTION
**Pull request by aarman-cloudera**
This allows a user to retrieve keytab that belong to other users
as long as they have the given right. This type of authorization
check doesn't seem to fit the existing authorization annotations
so I've instead put the logic in an AuthorizationUtil class.

### CODE REVIEW

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import javax.inject.Inject;
+
+@Component
+public class AuthorizationUtil {
+
+    @Inject
+    private GrpcUmsClient grpcUmsClient;
+
+    /**
+     * An authorization check that verifies that a calling actor can perform an operation against a
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: I would rephrase a bit saying that the check checks that the calling actor is either perform an action against themselves or have the right.

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+     * target user. The calling actor only requires the right when the target user is different.
+     */
+    public void checkCallerIsSelfOrHasRight(String actorCrnStr, String targetUserCrnStr, String requiredRight) {
+        Crn actorCrn = Crn.safeFromString(actorCrnStr);
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: assert the parameters?

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        Crn targetUserCrn = Crn.safeFromString(targetUserCrnStr);
+        if (actorCrn.equals(targetUserCrn)) {
+            return;
+        }
+        if (!actorCrn.getAccountId().equals(targetUserCrn.getAccountId())) {
+            throw new AccessDeniedException("Unauthorized to run this operation in a different account");
+        }
+        if (!grpcUmsClient.checkRight(GrpcUmsClient.INTERNAL_ACTOR_CRN, actorCrn.toString(), requiredRight, MDCUtils.getRequestId())) {
```
Comment (giladwolff) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: I think that checkRight throws if the check fails.

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+public class AuthorizationUtil {
+
+    @Inject
+    private GrpcUmsClient grpcUmsClient;
+
+    /**
+     * An authorization check that verifies that a calling actor can perform an operation against a
+     * target user. The calling actor only requires the right when the target user is different.
+     */
+    public void checkCallerIsSelfOrHasRight(String actorCrnStr, String targetUserCrnStr, String requiredRight) {
```
Comment (horadla23) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: we have `UmsAuthorizationService` class, which has methods for checking rights through UMS, you should move this method to that class and maybe you can also reuse those existing methods

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtV1Controller.java [hunk 5]
```diff
@@ -43,6 +45,9 @@
     @Inject
     private UserKeytabService userKeytabService;
 
+    @Inject
+    private AuthorizationUtil authorizationUtil;
```
Comment (horadla23) [inline freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtV1Controller.java]: as mentioned, please move your logic to `UmsAuthorizationService` and use that

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java (hunk 2 above)
Comment (aarman-cloudera) [reply freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: The safeFromString does a null check for us.

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java (hunk 3 above)
Comment (aarman-cloudera) [reply freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: So the grpcUmsClient checks for exception and returns a boolean instead.

File: freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java (hunk 4 above)
Comment (aarman-cloudera) [reply freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/AuthorizationUtil.java]: Thanks! didn't know about that service. I've moved thee method over.
